src/controller/java/CHIPDeviceController-JNI.cpp

```diff
@@ -477,11 +478,17 @@ void HandleSimpleOperationComplete(ChipDeviceController * deviceController, void
 
 void HandleKeyExchange(ChipDeviceController * deviceController, Transport::PeerConnectionState * state, void * appReqState)
 {
-    JNIEnv * env;
+    HandleSimpleOperationComplete(deviceController, appReqState);
+}
 
-    sJVM->GetEnv((void **) &env, JNI_VERSION_1_6);
+void HandleConnectIp(ChipDeviceController * deviceController, Transport::PeerConnectionState * state, void * appReqState)
+{
+    // In 'connect IP' aka 'connect without secure pairing' scenario, this callback is invoked synchronously
+    // while still holding sStackLock mutex. Therefore, we must temporarily release the lock here.
```

Comment:
this seems very coupled (and easy to miss as we evolve code).

Could we somehow design around it, like having the app state have some flags or additional settings to facilitate a safe unlock? Or even a pthread_mutex_trylock. If it succeeds, then we 'unlock' at exit. If it fails, we unlock and relock at exit.

Reply:
Also, what are we trying to lock out with this lock, and why can't it happen when we momentarily unlock here?  Or if it _can_ happen then, why is that OK?

Would maknig `sStackLock` be `PTHREAD_MUTEX_RECURSIVE` be appropriate?

Reply:
I'm not sure about pthread_mutex_trylock solution. Even if it fails we don't know if it's because we're holding the lock or somebody else is, so it's difficult to make the right decision based on that. Or am I missing something?
Having the app state which would tell us whether it's OK to initate a new controller transaction would work though it does sound to me like reimplementing a recursive mutex. Unless you meant that we could pass some kind of state to the callback which would then know that it doesn't need to take the lock?

Reply:
@bzbarsky-apple see Michael's comments in #3120 regarding recursive mutex. The lock seems to limit number of CHIPDeviceController users to 1. The problem is that callback of connectDeviceWithoutSecureSession is called synchronously (before the function exits) and wants to use the controller to send a message/command.

Reply:
I would suggest to append an NoLock() suffix to a HandleConnectIP and HandleSimpleOperationComplete and make them not lock. Then HandleSimpleOperationComplete could be of the form:

```
  pthread_mutex_lock(&sStacklock);
  HandleSimpleOperationCompleteNoLock(....);
  pthread_mutex_unlock(&sStackLock);
```

or invert and add 'Locked' suffix to functions that lock.

Reply:
I see.  So the point is that `ConnectDevice` calls the new-connection callback asynchronously, but `ConnectDeviceWithoutSecurePairing` calls it synchronously?

How viable is it to make `ConnectDeviceWithoutSecurePairing` call the callback asynchronously?  That seems like it would make code using it easier to migrate to `ConnectDevice` if/when we want to do it, and also prevent issues like this....

Reply:
Don't the async cases also hold the lock, and therefore HandleSimpleOperationComplete can always unlock while calling Java?

I see two solutions here:
- [reentrancy allowed] Always unlock while calling Java, since it may make a reentrant call.
- [reentrancy disallowed] Don't unlock, and app defers commands that result from a comment until the original call completes.

The [reentrancy allowed] option puts a burden on the CHIP stack to be reentrant safe, while [reentrancy disallowed] puts the burden on the Application to avoid these calls.

Reply:
In particular, in the current setup you can have a situation where you call `ChipDeviceController::ConnectDeviceWithoutSecurePairing` and get a no-error return from it but in fact the state of the controller is `kState_NotInitialized` because someone snuck in during the callback, before the return from `ConnectDeviceWithoutSecurePairing`, and called `ChipDeviceController::Shutdown` or some such.  With the async callback, that would not happen.  If someone calls `Shutdown` before the callback fires, it will just never fire, so no attempt would be made to use an invalid-state controller.

Reply:
Yes, we've got some JNI code which does:
```
pthread_mutex_lock(&sStacklock);
ConnectDeviceWithoutSecurePairing(..., HandleConnectIP, ...)
pthread_mutex_lock(&sStacklock);

And then HandleConnectIP calls (indirectly) another JNI piece which does something similar:
```
pthread_mutex_lock(&sStackLock);
SendMessage((void *) "SendMessage", buffer);
pthread_mutex_unlock(&sStackLock);
```
In case of asynchronous ConnectDevice unlocking in HandleSimpleOperationComplete isn't needed. Also releasing the lock **always** in HandleSimpleOperationComplete wouldn't be correct since the function is only called on a successful operation.

Reply:
> Don't the async cases also hold the lock

Not sure whether we're talking about the same thing, but the ordering in the async cases is as follows:
1. Take lock.
2. Call `ConnectDevice`.
3. Release lock.
4. Callback is called.
and the callback itself does not take the lock (though the Java code it calls can presumably call into other methods that do take the lock, around their `ChipDeviceController` manipulation).

So if callbacks from CHIP to the app are always async, we are effectively in a [reentrancy disallowed] case, but it's CHIP itself that does the deferring, removing that burden from the app.

I should also note that `ConnectDeviceWithoutSecurePairing` is probably not an API that actual apps will use much, if at all, hopefully.  But it's pretty useful for testing purposes.

Reply:
@bzbarsky-apple, probably making ConnectDeviceWithoutSecurePairing would be the best solution. The question is if it's worth putting effort if it's only a temporary code. Since CHIPDeviceController is meant to be highly portable code we would need to use some primitive already available in CHIP, like SystemTimer to schedule the call, right?

Reply:
Ok. I can do it. @mspang, @andy31415 would you approve such solution?

Reply:
> > Don't the async cases also hold the lock
> 
> Not sure whether we're talking about the same thing, but the ordering in the async cases is as follows:
> 
> 1. Take lock.
> 2. Call `ConnectDevice`.
> 3. Release lock.
> 4. Callback is called.
>    and the callback itself does not take the lock (though the Java code it calls can presumably call into other methods that do take the lock, around their `ChipDeviceController` manipulation).

The callback doesn't take the lock, but doesn't the IO thread hold it already?

https://github.com/project-chip/connectedhomeip/blob/fc4fec635cff1ac9585ffaaff223f0946ceb4b74/src/controller/java/CHIPDeviceController-JNI.cpp#L863

We only have the one lock. If we're not holding the lock, it's not safe to mutate CHIP data structures, including timers. On timer expiry, the first thing that happens is taking the lock.

> 
> So if callbacks from CHIP to the app are always async, we are effectively in a [reentrancy disallowed] case, but it's CHIP itself that does the deferring, removing that burden from the app.
> 
> I should also note that `ConnectDeviceWithoutSecurePairing` is probably not an API that actual apps will use much, if at all, hopefully. But it's pretty useful for testing purposes.

Reply:
> Ok. I can do it. @mspang, @andy31415 would you approve such solution?

I'm questioning the premise and I don't know that there is any difference between ConnectDeviceWithoutSecurePairing and ConnectDevice here.

Reply:
> but doesn't the IO thread hold it already?

Ah, I see.  In that case it seems like callbacks are always called with the lock held, and the principled thing for them to do is to always temporarily unlock and then re-lock?  It'd be nicer if we could drop the lock before the callbacks are called, but given the setup that doesn't seem viable.  And this seems like an issue for `ConnectDevice` as well, not just `ConnnectDeviceWithoutSecurePairing`, in that case?

Reply:
Yes, I think it's the case that any reentrant call inside a callback is going to deadlock and there is no difference between the various entry points in this respect.

We can release the lock before making the callback, or inside the callback, to allow reentrancy.

Or we can remove the re-entrancy as supporting it can be error prone.

My immediate suggestion is to always drop the lock in CHIPDeviceController-JNI.cpp, which allows the re-entrancy and fixes the regression, unblocking the app users & developers.

Reply:
Do we expect CHIP apps to be multi-threaded often?  If so, it might be worth trying to push down the unlocking into CHIP (e.g. by having the lock you take be CHIP-defined and either having `HandleSelectResult` have out-of-band knowledge of it or having `HandleSelectResult` take a reference to it, and then pass it down to things like `Timer::HandleComplete` so it can unlock around calling the callback.  The key part if we do this is to not introduce locking/unlocking in the single-threaded case, right?

If the multi-threaded case is rare, maybe pushing this complexity onto the apps is ok....

Reply:
(longer term, I think disallowing reentrancy may be a better option, or at least auditing the code to see how well we're currently handling reentrancy. So we might consider adding a TODO about that)

Reply:
The POSIX CHIP port is actually inherently multithreaded, because of the IO thread. It currently already forces the application to be multithreaded.

Reply:
Then long-term it seems like we really should push this complexity down into CHIP.  Short-term, I agree that we should just add unlocking in all the `CHIPDeviceController-JNI.cpp` callbacks for now and file followups to sort out a better approach.